Remove jsrsasign

linjie997 opened the pull request:
This is still a WIP, but in my local environment it works well.
I have tried login with code flow and refresh tokens, working as expected.
Some unit tests still need to be fixed.

Related to https://github.com/damienbod/angular-auth-oidc-client/issues/506 and https://github.com/damienbod/angular-auth-oidc-client/issues/26.
This would be a pretty big change in the library since we could remove the big dependency from jsrsasign, so some feedback/opinion would be appreciated since I'm not 100% sure about the changes so far.

damienbod commented:
@linjie997 Thank for this, I would need to do a lot of testing for this, not sure if this will work. I will have to setup all the supported keys on my test IDP. Won't have time for the near future, got a lot happening in the next week. Will get around to this then.

We would need to test this against all the RSA and ECDSA certificates.

Greetings Damien

linjie997 commented:
Sure, let's keep in touch!

d-koppenhagen commented:
Hey! Is there any roadmap when this feature/refactoring will be released? 
Will it be a breaking change, as the API isn't implemented in older (legacy) browsers? 
Probably a hint for using a polyfill would be good wouldn't it?

linjie997 commented:
Hello @d-koppenhagen, from my perspective it should be working and just need to be tested as damien said.
About the legacy browsers (I guess you mean IE), this implementation uses [Crypto.subtle](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/subtle), so if IE support is needed maybe [this polyfill](https://github.com/PeculiarVentures/webcrypto-liner) might work.

damienbod commented:
I hope to get around to testing this soon, I have a few different key types to test this.

FabianGosebrink commented:
Hey, this is great work. Can you fix the unit tests? Thanks

linjie997 commented:
I'll try to fix them as soon as I find some time, I'm currently on vacation :)

FabianGosebrink commented:
That is absolutely more important than coding! Enjoy the vacation 👋 ☀⛱🌊

linjie997 commented:
@FabianGosebrink ehy, unit tests should have been all fixed.
I have been a bit busy with vacation and new job so it took a while :)

FabianGosebrink commented:
Thanks for this. Can you fix the linting please?

linjie997 commented:
Linting should be fine now!

damienbod commented:
Hi @linjie997 Let me know if I can help, I will invest time here, see if I can solve this as well,

Greetings Damien

linjie997 commented:
Hi @damienbod, if you have time to take a look into it and give some help I would be really grateful! Unfortunately I have been very busy in RL

linjie997 commented:
@FabianGosebrink Hello Fabian, thank you for the code review.
I think you saw an older version of the code, with the latest commit I moved to observables instead of promises

FabianGosebrink commented:
Oh I am sorry! I will review again tomorrow and use the latest code then. Thanks for your effort! Would love to get this merged. @damienbod is currently testing iirc. Thanks 👍

linjie997 commented:
Greetings!
I finally had some time and I pushed some fix for various keys and conflicts from main branch.
My tests passed and it would be great to have some feedback from you @damienbod.
Sorry for taking so long and hope to hear you soon.

damienbod commented:
@linjie997 awesome, testing now, thanks

linjie997 commented:
Some unit test are failing at the moment, if the code itself looks good and keys test are passing I will work on the unit testing too, there are quite a bit to refactor :)

FabianGosebrink commented on projects/angular-auth-oidc-client/package.json:
```diff
@@ -9,7 +9,10 @@
     "tslib": "^2.3.0",
-    "jsrsasign-reduced": "^8.0.15"
+    "@angular-devkit/schematics": "^12.0.0",
+    "@schematics/angular": "^12.0.0",
```
remove

FabianGosebrink commented on projects/angular-auth-oidc-client/package.json:
```diff
@@ -9,7 +9,10 @@
-    "jsrsasign-reduced": "^8.0.15"
+    "@angular-devkit/schematics": "^12.0.0",
+    "@schematics/angular": "^12.0.0",
+    "@angular-devkit/core": "^12.0.0",
```
remove

FabianGosebrink commented on projects/angular-auth-oidc-client/ng-package.json:
```diff
@@ -2,7 +2,10 @@
   "lib": {
-    "entryFile": "src/public-api.ts"
+    "entryFile": "src/public-api.ts",
+    "umdModuleIds": {
```
remove the umdModuleIds please

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts:
```diff
@@ -21,21 +22,25 @@ export class StateValidationCallbackHandlerService {
   // STEP 4 All flows
 
   callbackStateValidation(callbackContext: CallbackContext, configId: string): Observable<CallbackContext> {
```
Change returntype to `: Observable<CallbackContext | Observable<never>>`

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts:
```diff
@@ -21,21 +22,25 @@ export class StateValidationCallbackHandlerService {
 
-      return throwError(() => new Error(errorMessage));
-    }
+          throwError(errorMessage);
```
remove this throwError as it is a merge mistake :)

FabianGosebrink commented:
I did a few comments because we could not build and run the lib. Please have a look, thanks!

linjie997 commented:
Updated, should be fine now

damienbod commented:
Thanks @linjie997 @FabianGosebrink tested the functionality on Firefox, Chrome and Edge

All EC and RSA work on Edge and Chrome

All RSA work on Firefox

 ECDsaSigningAlgorithm.ES256 and ES384 don't work on Firefox => this worked before, something small wrong here and only just for Firefox.

Maybe you see something, looking as well.

Greetings Damien

damienbod commented:
@linjie997 

I think this is the problem:

https://bugzilla.mozilla.org/show_bug.cgi?id=1718899

I think if the key EC, we need to remove the alg value so that firefox works, no idea how this effects the Chrome and the Edge implementations, testing now

would need some type of fix here:

```
  return from(this.cyptoObj.subtle.importKey('jwk', key, algorithm, false, ['verify'])).pipe(
      mergeMap((cryptoKey: CryptoKey) => {
        const signature: Uint8Array = base64url.parse(rawSignature, {loose: true});

        const algorithm: RsaHashedImportParams | EcdsaParams = this.getVerifyAlg(alg);
        return from(this.cyptoObj.subtle.verify(algorithm, cryptoKey, signature, new TextEncoder().encode(signingInput)));
      }),
      tap((isValid: boolean) => {
        if (!isValid) {
          this.loggerService.logWarning(configId, 'incorrect Signature, validation failed for id_token');
        }
      })
    );
  }
```

damienbod commented:
This fixes the Firefox bug

```
const agent = window.navigator.userAgent.toLowerCase();

if (agent.indexOf('firefox') > -1 && key.kty === 'EC') {
  key.alg = '';
}

```

linjie997 commented:
Great finding, I will update the code later with this addition

linjie997 commented:
Updated with Firefox hack.
Seems that everything is working good, I will dedicate some time in the coming days to unit test fixes :)

damienbod commented:
This works now, really awesome, massive thanks!

All we need to now is a code clean up and code quality. We probably could rename the **JsrsAsignReducedService** service. And the file names to **jsrsasign-reduced.service.**

could you remove this from the package.json : "jsrsasign-reduced": "^8.0.15",

@FabianGosebrink we will release this as version 13.1.0 (before any other PRs , good?)

Greetings Damien

FabianGosebrink commented:
Yes, if the code builds and the tests are running I will do a review.

linjie997 commented:
Greetings. I have fixed the majority of unit test, there is still one failing but I'm not sure how to fix it.
It's in the TokenValidationService, do you perhaps have time to take a look at it?
A second point would be how to rename the JsrsAsignReducedService, since it is not using JsrsAsign anymore.

Thanks!

FabianGosebrink commented:
I just improved the tests a bit and saw that you are using `async` for the tests. In Angular, there is no `async` to wait for a test to complete all its asynchronous operations anymore. It is all `waitForAsync`. The tests now run, but the one fails:

```
TokenValidationService validateIdTokenAtHash returns true if sha is sha256 and generated hash equals atHash param FAILED
        Error: Expected false to equal true.
``` 

I don't know about the encryption, so maybe you can have a look, or @damienbod can help. 

@linjie997 Can you please replace all `async` in the tests with `waitForAsync`? Thanks.

linjie997 commented:
Hello,
I removed all unused `async`.
Abou the failing unit test, I think it was a merge mistake and I reverted to the old `atHash` value. Now the test is passing :)

FabianGosebrink commented:
Please do Singlequotes everywhere, no double quotes.

FabianGosebrink commented on .prettierrc:
```diff
@@ -1,5 +1,6 @@
   "trailingComma": "es5",
-  "printWidth": 140
+  "printWidth": 140,
+  "jsxSingleQuote": true
```
We don't need this, do we?

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts:
```diff
@@ -21,21 +22,23 @@ export class StateValidationCallbackHandlerService {
+          this.loggerService.logWarning(configId, errorMessage);
+          this.resetAuthDataService.resetAuthorizationData(configId);
+          this.publishUnauthorizedState(callbackContext.validationResult, callbackContext.isRenewProcess);
+          throw new Error(errorMessage);
```
Please use rxjs operator: return throwError(() => new Error(errorMessage));

linjie997 replied:
Using `return throwError(() => new Error(errorMessage));` would make the method return `Observable<CallbackContext | Observable<never>>` and cause errors in `FlowsService`, do you think it's the best approach? We would need some checks and rewrite in FlowsService.

FabianGosebrink replied:
Nice catch. I skipped the `map` operator for some reason. Okay then, `throw new Error()` is okay then, just add a new line before.

linjie997 replied:
Done :)

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/iframe/refresh-session-iframe.service.spec.ts:
```diff
@@ -34,46 +34,46 @@ describe('RefreshSessionIframeService ', () => {
-      })
-    );
-  });
+  // describe('refreshSessionWithIframe', () => {
```
Can this be deleted? Or do we need it and it needs to be taken a look at?

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/iframe/refresh-session-iframe.service.ts:
```diff
@@ -21,9 +22,13 @@ export class RefreshSessionIframeService {
   refreshSessionWithIframe(configId: string, customParams?: { [key: string]: string | number | boolean }): Observable<boolean> {
     this.loggerService.logDebug(configId, 'BEGIN refresh session Authorize Iframe renew');
-    const url = this.urlService.getRefreshSessionSilentRenewUrl(configId, customParams);
+    const urlPromise = this.urlService.getRefreshSessionSilentRenewUrl(configId, customParams);
```
This returns an observable, doesnt it?

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/login/par/par.service.ts:
```diff
@@ -32,23 +32,25 @@ export class ParService {
-        this.loggerService.logError(configId, errorMessage, error);
-
-        return throwError(() => new Error(errorMessage));
+    return from(this.urlService.createBodyForParCodeFlowRequest(configId, customParams)).pipe(
```
no `from` needed, `createBodyForParCodeFlowRequest` returns an observable.

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/login/popup/popup-login.service.spec.ts:
```diff
@@ -1,24 +1,24 @@
-import { PopUpServiceMock } from './popup.service-mock';
-
-describe('PopUpLoginService', () => {
+import { CommonModule } from "@angular/common";
```
Singlequotes

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/utils/tokenHelper/token-helper.service.ts:
```diff
@@ -17,6 +18,17 @@ export class TokenHelperService {
+    const header: string = this.getHeaderFromToken(token, encoded, configId);
+    const payload: string = this.getPayloadFromToken(token, encoded, configId);
+
+    return [header, payload].join('.');
```
I like this, good job!

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/utils/url/url.service.ts:
```diff
@@ -376,30 +384,36 @@ export class UrlService {
-      return null;
-    }
+        if (!silentRenewUrl) {
+          return null;
```
lets return an empty string here.

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/utils/url/url.service.ts:
```diff
@@ -436,16 +450,19 @@ export class UrlService {
+        this.loggerService.logError(configId, 'authWellKnownEndpoints is undefined');
 
-    return null;
+        return null;
```
lets return an empty string here

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/validation/token-validation.service.ts:
```diff
@@ -417,30 +481,34 @@ export class TokenValidationService {
-    return false;
+    return this.jsrsAsignReducedService.generateAtHash('' + accessToken, sha).pipe(
+      mergeMap((hash: string) => {
+        console.warn('hash', hash);
```
remove console.warn

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/utils/url/url.service-mock.ts:
```diff
@@ -44,8 +45,15 @@ export class UrlServiceMock {
-  createBodyForParCodeFlowRequest(customParamsRequest?: { [key: string]: string | number | boolean }): string {
-    return '';
+  createBodyForParCodeFlowRequest(customParamsRequest?: { [key: string]: string | number | boolean }): Promise<string> {
+    return Promise.resolve('');
```
Observable here

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/validation/token-validation.service-mock.ts:
```diff
@@ -31,16 +31,16 @@
 
-  validateIdTokenAtHash(accessToken: any, atHash: any, isCodeFlow: boolean): boolean {
-    return true;
+  validateIdTokenAtHash(accessToken: any, atHash: any, isCodeFlow: boolean): Promise<boolean> {
```
Observable

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/validation/token-validation.service-mock.ts:
```diff
@@ -31,16 +31,16 @@
 
-  validateSignatureIdToken(idToken: any, jwtkeys: any): boolean {
-    return true;
+  validateSignatureIdToken(idToken: any, jwtkeys: any): Promise<boolean> {
```
Observable

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/login/standard/standard-login.service.spec.ts:
```diff
@@ -1,20 +1,20 @@
-import { StandardLoginService } from './standard-login.service';
-
-describe('StandardLoginService', () => {
+import { fakeAsync, TestBed, tick, waitForAsync } from "@angular/core/testing";
```
Singlequotes all over

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/validation/state-validation.service.spec.ts:
```diff
@@ -101,14 +102,15 @@ describe('State Validation Service', () => {
   });
 
-  it('should return invalid result if validateStateFromHashCallback is false', () => {
+  it('should return invalid result if validateStateFromHashCallback is false', async () => {
```
waitForASync

FabianGosebrink commented:
Please do `waitForAsync` all over and Single quotes all over. Thanks!

linjie997 commented:
Should be good now, I checked the changes one more time

linjie997 commented:
Should be good now, I checked the changes one more time

FabianGosebrink commented on projects/angular-auth-oidc-client/src/lib/utils/url/url.service-mock.ts:
```diff
@@ -44,8 +46,15 @@ export class UrlServiceMock {
+    return of('');
+  }
+
+  private async createUrlCodeFlowWithSilentRenew(
```
No async, no Promise, just Observable here

FabianGosebrink commented:
Cool! We are getting there. Just one small comment. @damienbod can you do a last test run before merge?

damienbod commented:
I do a test then, thanks @linjie997 @FabianGosebrink

linjie997 commented:
Great, it took a while but I'm glad it worked out good at the end.
I'll wait for the last tests feedback!

damienbod commented:
@linjie997 Thanks for this, this was a huge PR and took so long. Thanks so much for you patience and effort

preparing the release now 13.1.0

Greetings Damien

linjie997 commented:
Great news, happy that I contribuited!!

FabianGosebrink commented:
Congratulations! Great one!

jeroenheijmans commented:
I'd been following this PR with great interest, and at the risk of posting a "+1" style comment I wanted to explicitly thank y'all for this hard work in creating, reviewing, and completing this PR! This work will serve the ecosystem, the devs, but even more so many consumers (and their data plans 😅) downstream. Great work!

damienbod commented:
Thanks @jeroenheijmans

FabianGosebrink commented:
Thanks @jeroenheijmans !

GabrielGil commented on projects/angular-auth-oidc-client/src/lib/validation/jsrsasign-reduced.service.ts:
```diff
@@ -1,20 +1,50 @@
-  generateCodeChallenge(codeVerifier: any): string {
-    const hash = KJUR.crypto.Util.hashString(codeVerifier, 'sha256');
-    const testData = hextob64u(hash);
+  private crypto: Crypto = window.crypto || (window as any).msCrypto; // for IE11
```
This (and the other global `window` calls) break support of Angular Universal (SSR).

The correct way to use window, is by injecting the document `@Inject(DOCUMENT) private readonly doc: Document,` and then using it `this.doc.defaultView`.

Also, since this version is for Angular 13, and the support  of Internet Explorer is deprecated by the Angular team, i don't think it makes sense to consider IE11.

FabianGosebrink replied:
That is absolutely correct. We already use the injection of the document in other places. We will fix that in a separate PR. Thanks for the hint!

GabrielGil replied:
No problem. We collaborate to help each other :) Thanks to you guys!

FabianGosebrink replied:
https://github.com/damienbod/angular-auth-oidc-client/issues/1316

GabrielGil reviewed (commented):
I suggest that `window` is never accessed from the global namespace, as it may not be available, for instance in the server, and the injection of the document should be used instead.

Refer to #913 for more.

Budno commented:
@FabianGosebrink Could you mark this as a breaking change? Because due to #1320 this ends the support for http.

FabianGosebrink commented:
Why is it ending the HTTP support? We still use HTTP to communicate.

DamienBowden commented:
@Budno  @FabianGosebrink yes https is required to use the browser APIs

Lexa-tech commented:
Hi @linjie997 , did you try the polyfill webcrypto-liner you mentionned here when using HTTP ?
https://github.com/damienbod/angular-auth-oidc-client/pull/1220#issuecomment-908551547

IlyaSurmay commented:
> Hi @linjie997 , did you try the polyfill webcrypto-liner you mentionned here when using HTTP ? [#1220 (comment)](https://github.com/damienbod/angular-auth-oidc-client/pull/1220#issuecomment-908551547)

I tried it and it didn't seem to work
<img width="742" alt="image" src="https://github.com/damienbod/angular-auth-oidc-client/assets/18414330/162e70e6-12ce-429a-814b-86206840c243">
adding asmcrypto library didn't help much, but the error message changed to `ERROR TypeError: asmCrypto.SHA256 is undefined`

desmondblue commented:
I went through the conversation and saw @linjie997  had pushed some fixes to version 13.x version.  Hence my question - Does that imply the breaking change that @Budno mentions "browser APIs will only use https" is starting from 13.x while the 12.x plugin may work fine without the crypto.digest issue?
